test-requirements.txt

```diff
@@ -6,6 +6,7 @@ flake8-builtins
 isort
 mypy
 pep8-naming
+protobuf<=3.20.*  # https://protobuf.dev/news/2022-05-06/#python-updates
```

Comment:
Where is this used?

Reply:
protobuf is imported indirectly through pytest_operator -> juju client.
If we do not pin the version, the final versions 4.x is installed and it has some breaking changes with the existing code.

```
 File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/pytest_operator/plugin.py", line 48, in <module>
    from juju.model import Model, Controller, websockets
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/juju/model.py", line 29, in <module>
    from .client import client, connector
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/juju/client/connector.py", line 7, in <module>
    import macaroonbakery.httpbakery as httpbakery
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/macaroonbakery/httpbakery/__init__.py", line 3, in <module>
    from ._client import (
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/macaroonbakery/httpbakery/_client.py", line 7, in <module>
    import macaroonbakery.bakery as bakery
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/macaroonbakery/bakery/__init__.py", line 75, in <module>
    from ._oven import (
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/macaroonbakery/bakery/_oven.py", line 29, in <module>
    from ._internal import id_pb2
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/macaroonbakery/bakery/_internal/id_pb2.py", line 35, in <module>
    _descriptor.FieldDescriptor(
  File "/home/gatici/charmed5g/sdcore-router-operator/.tox/unit/lib/python3.11/site-packages/google/protobuf/descriptor.py", line 553, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

```

Reply:
This issue is probably triggered by macaroonbakery 1.3.3 which is published yesterday.

Reply:
Got it. Personally I'm not a fan of ranges. How about we pin to a specific version here (i.e. last working)?

Reply:
Also... if it's `macaroonbakery` that causes the problem, can't we pin it instead of `protobuf`?

Reply:
Yes, this also fixes the issue. I am going to pin macaroonbakery to 1.3.2.